gpt_engineer/chat_to_files.py

```diff
@@ -160,3 +161,67 @@ def format_file_to_input(file_name: str, file_content: str) -> str:
     ```
     """
     return file_str
+
+
+def overwrite_files_with_edits(chat: str, dbs: DBs):
+    edits = parse_edits(chat)
+    apply_edits(edits, dbs.workspace)
+
+
+@dataclass
+class Edit:
+    filename: str
+    before: str
+    after: str
+
+
+def parse_edits(llm_response):
+    def parse_one_edit(lines):
+        HEAD = "<<<<<<< HEAD"
+        DIVIDER = "======="
+        UPDATE = ">>>>>>> updated"
+
+        filename = lines.pop(0)
+        text = "\n".join(lines)
+        splits = text.split(DIVIDER)
+        if len(splits) != 2:
+            raise ValueError(f"Could not parse following text as code edit: \n{text}")
+        before, after = splits
+
+        before = before.replace(HEAD, "").strip()
+        after = after.replace(UPDATE, "").strip()
+
+        return Edit(filename, before, after)
+
+    def parse_all_edits(txt):
+        edits = []
+        current_edit = []
+        in_fence = False
+
+        for line in txt.split("\n"):
+            if line.startswith("```") and in_fence:
+                edits.append(parse_one_edit(current_edit))
+                current_edit = []
+                in_fence = False
+                continue
+            elif line.startswith("```") and not in_fence:
+                in_fence = True
+                continue
+
+            if in_fence:
+                current_edit.append(line)
+
+        return edits
+
+    return parse_all_edits(llm_response)
+
+
+def apply_edits(edits: List[Edit], workspace: DB):
+    for edit in edits:
+        filename = edit.filename.replace("workspace/", "")
+        if edit.before == "":
+            workspace[filename] = edit.after  # new file
+        else:
+            workspace[filename] = workspace[filename].replace(
```

Comment:
We first need to check that there is only one match here I think.

And at least `logger.warn` if there are multiple matches.

Reply:
Agree & changed